mac/Keyman4MacIM/Keyman4MacIM/Privacy/PrivacyConsent.m

```diff
@@ -0,0 +1,244 @@
+ * permission, so that access is not explicitly needed as long as one of
+ * the others is provided first.
+ *
+ * Access this functionality through the shared instance. Call
+ * requestPrivacyAccess which will check whether access has been granted.
+ * If not, then a dialog will be presented to the user to inform them
+ * that Accessibility is required. Dismissing this dialog will trigger
+ * the API call to prompt the user for permission. However, the prompt
+ * from the sytem will only happen once.
+ * If the user was prompted earlier and did not grant access, they can
+ * still go to Sytem Preferences and enable it there. So it is useful to prompt
+ * with a dialog from Keyman at startup.
+ * Without Accessibility, the shift layer and the OSK do not work -- so Keyman
+ * is basically useless.
+ */
+
+#import "PrivacyConsent.h"
+
+@implementation PrivacyConsent
+
++ (PrivacyConsent *)shared
+{
+    static PrivacyConsent *shared = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+      shared = [[PrivacyConsent alloc] init];
+    });
+    return shared;
+}
+
+- (BOOL)checkAccessibility
+{
+  BOOL hasAccessibility = NO;
+
+  NSLog(@"checkAccessibility called");
+  
+  hasAccessibility = AXIsProcessTrusted();
+  NSLog(@"  hasAccessibility: %@",hasAccessibility ? @"YES" : @"NO");
+  return hasAccessibility;
+}
+
+- (void)requestPrivacyAccess:(void (^)(void))completionHandler
+{
+  _handler = completionHandler;
+  BOOL hasAccessibility = NO;
+
+  NSLog(@"requestPrivacyAccess called");
+  
+  // check if we already have accessibility
+  if (@available(macOS 10.15, *)) {
+    hasAccessibility = [self checkPostEventAccess];
+  } else {
+    hasAccessibility = [self checkAccessibility];
+  }
+  
+  if (hasAccessibility) {
+    NSLog(@"already have Accessibility: no need to make request.");
+  } else {
+    NSLog(@"do not have Accessibility, present Privacy Dialog");
+    [self showPrivacyDialog];
+  }
+}
+
+- (void)requestAccessibility
+{
+  NSLog(@"requestAccessibility called");
+
+  NSDictionary *options = @{(id)CFBridgingRelease(kAXTrustedCheckOptionPrompt): @YES};
+  BOOL hasAccessibility = AXIsProcessTrustedWithOptions((CFDictionaryRef)options);
+  NSLog(@"  hasAccessibility: %@",hasAccessibility ? @"YES" : @"NO, requesting...");
+}
+
+- (NSWindowController *)privacyDialog {
+  NSLog(@"privacyDialog getter called");
+  if (!_privacyDialog) {
+        _privacyDialog = [[PrivacyWindowController alloc] initWithWindowNibName:@"PrivacyWindowController"];
+    NSLog(@"privacyDialog created");
+    [self configureDialogForOsVersion];
+  }
+
+  return _privacyDialog;
+}
+
+- (void)configureDialogForOsVersion {
+
+  if (@available(macOS 10.15, *)) {
+    [self configureDialogForCatalinaAndLater];
+  } else {
+    [self configureDialogForPreCatalina];
+  }
+}
+
+- (void)configureDialogForPreCatalina {
+  void (^consentPrompt)(void) = ^(void) {
+    NSLog(@"block executed!");
```

Comment:
We probably don't want quite so many logging messages in production code -- most of these won't be helpful in the future.